source/opt/allocator.h

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
```

Comment:
Did you look at importing something like partition alloc instead of a custom one?

Reply:
You mean https://chromium.googlesource.com/chromium/src/+/dcc13470a/third_party/WebKit/Source/wtf/PartitionAlloc.md?

IIRC, we previously discussed the possibility of leveraging some existing allocator. But it's an additional dependency; the external allocator has its own compiler requirements (yeah, VS2013). Copying in the code will have license problem; even if possible, we may end up pulling in too many source files. So we didn't do that.

I'm not quite happy of writing an allocator by myself, too. But it seems it's a more viable approach. Happy to explore alternatives for sure.

Reply:
There is one in base/ as well: https://chromium.googlesource.com/chromium/src/+/master/base/allocator/

I'm just concerned because writing an allocator is .... hard ... and can have security implications.

Reply:
That's a lot of source files... I'm not quite sure, but is it easily extractable and importable here?

Yep. I totally agree. That's why I added a lot of reviwers. :) The one I have here is basically referencing what Clang `BumpPtrAllocator` is doing.

Reply:
Yea, we pulled it into pdfium: https://pdfium.googlesource.com/pdfium/+/master/third_party/base/allocator/partition_allocator there are bits in there and in third_party/base that it needs. But, yea, it's large .....